torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+from ..modules.module import Module
+from ..parameter import Parameter
+import functools
+import inspect
+
+
+class Parametrization(Module):
+    r"""A kind of Module that parametrizes a Parameter or a buffer in terms
+    of its forward function
+
+    After registering a Parametrization ``param`` on a tensor ``t`` within a module
+    ``module`` with :meth:`#register_parametrization`, ``module.t`` will be turn
+    into a property, and will return ``param(t)``, where ``t`` is the previous tensor
+
+    After registering a Parametrization, the parametrization will be accessible under
+    ``module.t_param``. The parametrized and unparametrized tensors may be accessed
+    through ``param.evaluate()`` and ``param.original()``
+
+    Before a parametrization is registered, it may be chained to others to compose them
+    via the ``chain`` method. A parametrization may be evaluated on a tensor ``t`` by
+    ``param.evaluate(t)``, even when it has not been registered
+
+    Parametrizations may be registered on parameters, buffers, and already parametrized
+    tensors
+
+        # Defining several parametrizations on the same tensor
+        module = nn.Linear(5,3)
+        p1 = MyParametrization()
+        p2 = MyParametrization()
+        # Register p1 on the Parameter "weight"
+        register_parametrization(module, "weight", p1)
+        # Register p2 on the parametrized Parameter "weight"
+        register_parametrization(module, "weight", p2)
+        assert(module.weight == p2(p1(module.weight_param.original())))
+
+    We can also first compose them and then register them. The following code is
+    equivalent to the previous one
+
+        module = nn.Linear(5,3)
+        # Now p2.evaluate computes p1 and then p2
+        p2.chain(p1)
+        # Register the chained parametrization
+        register_parametrization(module, "weight", p2)
+        assert(module.weight == p2(p1(module.weight_param.original())))
+
+
+    Parametrized parameters and buffers have an in-built caching system via the context
+    manager :class:``cached`` and the decorator :class: ``cached_method``
+    """
+
+    def evaluate(self, tensor=None):
+        r""" It evaluates the current chain of parametrizations on a given tensor.
+
+        If the tensor is None, and the parametrization has been registered on a module,
+        it evaluates the parametrizaiton on the tensor on which it was registered
+
+        Args:
+            tensor (torch.Tensor): the at which to evaluate the parametrization chain.
+            Default: None
+        """
+        # TODO Write docs
+        if self.is_chained():
+            tensor = self.orig_param.evaluate(tensor)
+        else:
+            if tensor is None:
+                try:
+                    tensor = self.orig
```

Comment:
nit: `orig` is not a lot shorter than `original` and the latter reads much better. Please don't shorten the names too much.

Reply:
I took the wording from the current implementation of the pruning module, but I agree. I'll change it.

Reply:
Nevermind, I kept the name `orig`, because you also have the `original()` method that returns the original parameter from a parametrization chain. Should I change orig to `_original` or so?

Reply:
Yeah let's do `_original` then.